File: concepts/Refactoring/PHP/PHP Refactoring Best Practices/4Couplers/3ExcessiveExposure/solution/factory/src/handlers/SimpleCurrencyConverter.php
<?php 
namespace app\handlers;

use Exception;

class SimpleCurrencyConverter
{
  private $currencyTo;
  private $httpHelper;
  private function __construct($currencyTo, HttpHelper $httpHelper)
  {
    $this->currencyTo = $currencyTo;
    $this->httpHelper = $httpHelper;
  }

  public function convert($price)
  {
    if ($this->currencyTo === "EUR") {
      return $price * 0.9;
    } else if ($this->currencyTo === "CAD") {
      return $price * 1.35;
    } else
      throw new Exception("Unrecognized currency: $this->currencyTo");
  }


  public function convertWithWebService($price)
  {
    $rates = $this->httpHelper->getCurrencyRate("USD");
    if ($this->currencyTo === "EUR") {
      return $price * $this->getRate($rates, "EUR");
    } else if ($this->currencyTo === "CAD") {
      return $price * $this->getRate($rates, "CAD");;
    } else
      throw new Exception("Unrecognized currency: $this->currencyTo");
  }
  private function getRate(string $rates, string $currencyTo)
  {
    $data = json_decode($rates);
    return $data->rates->$currencyTo;
  }
  //new requirement: print rates : sol1 & 2
  public function printRates(string $currencyTo)
  {
    $rates = $this->httpHelper->getCurrencyRate("USD");
    if ($currencyTo === "EUR") {
      echo $this->getRate($rates, "EUR");
    } else if ($currencyTo === "CAD") {
      echo $this->getRate($rates, "CAD");;
    } else
      throw new Exception("Unrecognized currency: $this->currencyTo");
  }
  public function printRates1(string $currencyTo)
  {
    $rates = $this->httpHelper->getCurrencyRate("USD");
    echo $this->getRate($rates, $currencyTo);

  }

  public static function getInstance($currencyTo): SimpleCurrencyConverter
  {
    return new SimpleCurrencyConverter($currencyTo, new HttpHelper());
  }
}

File: software-engineering-concepts/Refactoring/Java/Java Refactoring Best Practices/ChangePreventers/DivergentChange/BaseProject/src/main/java/com/onlineshop/handlers/SimpleCurrencyConverter.java
package com.onlineshop.handlers;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

import static java.net.http.HttpClient.newBuilder;

public class SimpleCurrencyConverter {

    private HttpHelper httpHelper = new HttpHelper();
    private String currencyTo;

    public SimpleCurrencyConverter(){}


    public SimpleCurrencyConverter(String currencyTo){
        this.currencyTo = currencyTo;
    }

    public double convert(double price){
        if(currencyTo.equalsIgnoreCase("EUR")){
            return price * 0.9;
        } else if (currencyTo.equalsIgnoreCase("CAD")){
            return price* 1.35;
        } else {
            throw new IllegalArgumentException("Unrecognized currency : " + currencyTo);
        }
    }

    public double convertWithWebService(double price){

        String rates = httpHelper.getCurrencyRates("USD");
        if(currencyTo.equalsIgnoreCase("EUR")){
            return price * getRate(rates, "EUR");

        } else if (currencyTo.equalsIgnoreCase("CAD")){
            return price * getRate(rates, "CAD");
        } else {
            throw new IllegalArgumentException("Unrecognized currency : " + currencyTo);
        }
    }


    private double getRate(String rates, String currencyTo) {
        return Double.valueOf(rates.substring(rates.lastIndexOf(currencyTo)).substring(5,9));
    }



    // NEW REQUIREMENT TO PRINT RATE
    public void printUSDRate(String  currencyTo){
        String rates = httpHelper.getCurrencyRates("USD");
        System.out.println(getRate(rates, currencyTo));
    }


}
